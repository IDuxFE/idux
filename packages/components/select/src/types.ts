/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { AbstractControl, ValidateStatus } from '@idux/cdk/forms'
import type { PortalTargetType } from '@idux/cdk/portal'
import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { EmptyProps } from '@idux/components/empty'
import type { FormSize } from '@idux/components/form'
import type { DefineComponent, FunctionalComponent, HTMLAttributes, PropType, VNode, VNodeChild } from 'vue'

export const selectPanelProps = {
  activeValue: { type: [String, Number, Symbol] as PropType<VKey>, default: undefined },
  selectedKeys: { type: Array as PropType<VKey[]>, default: undefined },
  childrenKey: { type: String, default: undefined },

  customAdditional: { type: Function as PropType<SelectCustomAdditional>, default: undefined },
  dataSource: { type: Array as PropType<SelectData[]>, default: undefined },
  empty: { type: [String, Object] as PropType<string | EmptyProps>, default: undefined },
  getKey: { type: [String, Function] as PropType<string | ((data: SelectData<any>) => any)>, default: undefined },
  labelKey: { type: String, default: undefined },
  multiple: { type: Boolean, default: false },
  multipleLimit: { type: Number, default: Number.MAX_SAFE_INTEGER },
  virtual: { type: Boolean, default: false },

  // events
  'onUpdate:activeValue': [Function, Array] as PropType<MaybeArray<(value: any) => void>>,
  onOptionClick: [Function, Array] as PropType<MaybeArray<(option: SelectData<any>, evt: MouseEvent) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: SelectData<any>[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,

  // private
  _virtualScrollHeight: { type: Number, default: 256 },
  _virtualScrollItemHeight: { type: Number, default: 32 },
} as const

export const selectProps = {
  control: { type: [String, Number, Object] as PropType<string | number | AbstractControl>, default: undefined },
  value: { type: [String, Number, Symbol, Array] as PropType<MaybeArray<VKey>>, default: undefined },
  open: { type: Boolean, default: undefined },

  allowInput: { type: Boolean, default: false },
  autocomplete: { type: String, default: 'off' },
  autofocus: { type: Boolean, default: false },
  borderless: { type: Boolean, default: undefined },
  childrenKey: { type: String, default: undefined },
  clearable: { type: Boolean, default: false },
  clearIcon: { type: String, default: undefined },
  customAdditional: { type: Object as PropType<SelectCustomAdditional>, default: undefined },
  dataSource: { type: Array as PropType<SelectData[]>, default: undefined },
  disabled: { type: Boolean, default: false },
  empty: { type: [String, Object] as PropType<string | EmptyProps>, default: undefined },
  getKey: { type: [String, Function] as PropType<string | ((data: SelectData) => VKey)>, default: undefined },
  labelKey: { type: String, default: undefined },
  /**
   * @deprecated please use `maxLabel` instead'
   */
  maxLabelCount: { type: [Number, String] as PropType<number | 'responsive'>, default: undefined },
  maxLabel: { type: [Number, String] as PropType<number | 'responsive'>, default: Number.MAX_SAFE_INTEGER },
  multiple: { type: Boolean, default: false },
  multipleLimit: { type: Number, default: Number.MAX_SAFE_INTEGER },
  /**
   * @deprecated please use `dataSource` instead'
   */
  offset: Array as unknown as PropType<[number, number]>,
  options: { type: Array as PropType<SelectData[]>, default: undefined },
  overlayClassName: { type: String, default: undefined },
  overlayContainer: {
    type: [String, HTMLElement, Function] as PropType<PortalTargetType>,
    default: undefined,
  },
  overlayMatchWidth: { type: Boolean, default: undefined },
  overlayRender: { type: Function as PropType<(children: VNode[]) => VNodeChild>, default: undefined },
  placeholder: { type: String, default: undefined },
  readonly: { type: Boolean, default: false },
  searchable: { type: [Boolean, String] as PropType<boolean | 'overlay'>, default: false },
  /**
   * @deprecated please use `searchFn` instead'
   */
  searchFilter: { type: [Boolean, Function] as PropType<boolean | SelectSearchFn>, default: undefined },
  searchFn: { type: [Boolean, Function] as PropType<boolean | SelectSearchFn>, default: true },
  size: { type: String as PropType<FormSize>, default: undefined },
  status: String as PropType<ValidateStatus>,
  suffix: { type: String, default: undefined },
  /**
   * @deprecated please use `overlayContainer` instead'
   */
  target: {
    type: [String, HTMLElement, Function] as PropType<PortalTargetType>,
    default: undefined,
  },
  /**
   * @deprecated please use `getKey` instead'
   */
  valueKey: { type: String, default: undefined },
  virtual: { type: Boolean, default: false },

  // events
  'onUpdate:value': [Function, Array] as PropType<MaybeArray<(value: any) => void>>,
  'onUpdate:open': [Function, Array] as PropType<MaybeArray<(opened: boolean) => void>>,
  onChange: [Function, Array] as PropType<MaybeArray<(value: any, oldValue: any) => void>>,
  onClear: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onSearch: [Function, Array] as PropType<MaybeArray<(searchText: string) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: SelectData<any>[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,

  // private
  overlayHeight: { type: Number, default: 256 },
  overlayItemHeight: { type: Number, default: 32 },
} as const

export type SelectPanelProps = ExtractInnerPropTypes<typeof selectPanelProps>
export type SelectPanelPublicProps = Omit<
  ExtractPublicPropTypes<typeof selectPanelProps>,
  '_virtualScrollHeight' | '_virtualScrollItemHeight'
>
export interface SelectPanelBindings {
  scrollTo: VirtualScrollToFn
  changeActiveIndex: (offset: number) => void
}
export type SelectPanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof SelectPanelPublicProps> & SelectPanelPublicProps,
  SelectPanelBindings
>
export type SelectPanelInstance = InstanceType<DefineComponent<SelectPanelProps, SelectPanelBindings>>

export type SelectProps = ExtractInnerPropTypes<typeof selectProps>
export type SelectPublicProps = Omit<
  ExtractPublicPropTypes<typeof selectProps>,
  'maxLabelCount' | 'options' | 'searchFilter' | 'overlayHeight' | 'overlayItemHeight' | 'target' | 'valueKey'
>
export interface SelectBindings {
  blur: () => void
  focus: (options?: FocusOptions) => void
  scrollTo: VirtualScrollToFn
}
export type SelectComponent = DefineComponent<
  Omit<HTMLAttributes, keyof SelectPublicProps> & SelectPublicProps,
  SelectBindings
>
export type SelectInstance = InstanceType<DefineComponent<SelectProps, SelectBindings>>

export type SelectCustomAdditional = (options: {
  data: SelectData<any>
  index: number
}) => Record<string, any> | undefined

export interface SelectOptionProps<K = VKey> {
  /**
   * @deprecated please use `customAdditional` instead'
   */
  additional?: {
    class?: any
    style?: any
    [key: string]: unknown
  }

  disabled?: boolean
  key?: K
  label?: string | number

  customLabel?: string | ((data: SelectOptionProps<K>) => VNodeChild)

  [key: string]: any
}
export type SelectOptionPublicProps = Omit<SelectOptionProps, 'additional'>
export type SelectOptionComponent = FunctionalComponent<
  Omit<HTMLAttributes, keyof SelectOptionPublicProps> & SelectOptionPublicProps
>

export interface SelectOptionGroupProps<K = VKey> {
  /**
   * @deprecated please use `customAdditional` instead'
   */
  additional?: {
    class?: any
    style?: any
    [key: string]: unknown
  }

  children?: SelectOptionProps<K>[]
  key?: K
  label?: string | number

  customLabel?: string | ((data: SelectOptionGroupProps<K>) => VNodeChild)

  [key: string]: any
}
export type SelectOptionGroupPublicProps = Omit<SelectOptionGroupProps, 'additional'>
export type SelectOptionGroupComponent = FunctionalComponent<
  Omit<HTMLAttributes, keyof SelectOptionGroupPublicProps> & SelectOptionGroupPublicProps
>

export type SelectData<K = VKey> = SelectOptionProps<K> | SelectOptionGroupProps<K>

export type SelectSearchFn = (data: SelectData<any>, searchText: string) => boolean

export const optionProps = {
  disabled: { type: Boolean, default: false },
  index: { type: Number, required: true },
  label: { type: [String, Number] as PropType<string | number>, default: undefined },
  rawData: { type: Object as PropType<SelectOptionProps>, required: true },
  parentKey: { type: [String, Number, Symbol] as PropType<VKey>, default: undefined },
} as const

export type OptionProps = ExtractInnerPropTypes<typeof optionProps>

export const optionGroupProps = {
  label: { type: [String, Number] as PropType<string | number>, default: undefined },
  index: { type: Number, required: true },
  rawData: { type: Object as PropType<SelectOptionGroupProps>, required: true },
} as const

export type OptionGroupProps = ExtractInnerPropTypes<typeof optionGroupProps>
