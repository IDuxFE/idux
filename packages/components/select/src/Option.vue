<template>
  <div v-show="visible" :class="classes" @mouseenter="onMouseEnter" @click.stop="onClick">
    <span v-if="showCheckbox" class="ix-option-checkbox">
      <ix-checkbox :checked="selected" :disabled="disabled" readonly />
    </span>
    <span class="ix-option-label">
      <slot>{{ label }}</slot>
    </span>
  </div>
</template>
<script lang="ts">
import type { OptionProps, SelectFilterFn } from './types'

import { computed, defineComponent, inject, nextTick, onUnmounted, watch } from 'vue'
import { isFunction } from '@idux/cdk/utils'
import { IxCheckbox } from '@idux/components/checkbox'
import { optionPropsDef } from './types'
import { selectToken, visibleChangeToken } from './utils'

const defaultFilterFn: SelectFilterFn = (value: string, option: OptionProps) => {
  return option.label.toLowerCase().includes(value.toLowerCase())
}

export default defineComponent({
  name: 'IxOption',
  components: { IxCheckbox },
  props: optionPropsDef,
  setup(props: OptionProps) {
    const {
      selectProps,
      inputValue,
      activatedValue,
      activateHandler,
      selectedValue,
      selectHandler,
      selectOptionHandler,
    } = inject(selectToken) || {}

    const visible = computed(() => {
      const searchValue = inputValue?.value
      const filterOption = selectProps?.filterOption
      if (searchValue && filterOption) {
        const filterFn = (isFunction(filterOption) ? filterOption : defaultFilterFn) as SelectFilterFn
        return filterFn(searchValue, props)
      } else {
        return true
      }
    })

    const active = computed(() => selectProps?.compareWith(activatedValue?.value, props.value))
    const selected = computed(() => !!selectedValue?.value.some(item => selectProps?.compareWith(item, props.value)))

    watch(selected, value => nextTick(() => selectOptionHandler?.(value, props)), { immediate: visible.value })

    const classes = computed(() => {
      const _disabled = props.disabled
      return {
        'ix-option': true,
        'ix-option-disabled': _disabled,
        'ix-option-active': !_disabled && active.value,
        'ix-option-selected': !_disabled && selected.value,
      }
    })

    const showCheckbox = computed(() => selectProps?.multiple)

    const onMouseEnter = () => {
      if (!props.disabled) {
        activateHandler?.(props.value)
      }
    }

    const onClick = () => {
      if (!props.disabled) {
        selectHandler?.(props.value)
      }
    }

    const visibleChange = inject(visibleChangeToken)
    watch(visible, value => nextTick(() => visibleChange?.(value)), { immediate: true })
    onUnmounted(() => {
      if (visible.value) {
        visibleChange?.(false)
      }
    })

    return { visible, classes, showCheckbox, selected, onMouseEnter, onClick }
  },
})
</script>
