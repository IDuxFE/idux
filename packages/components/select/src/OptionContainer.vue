<template>
  <div class="ix-option-container">
    <template v-for="option in options" :key="option.value">
      <ix-option-group v-if="option.children" :label="option.label" :options="option.children" />
      <ix-option v-else :disabled="option.disabled" :label="option.label" :value="option.value" />
    </template>
    <slot v-if="!options.length" />
    <div v-show="isEmpty" class="ix-option-container-empty">
      <slot name="empty">
        <ix-empty :description="empty" />
      </slot>
    </div>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, provide, ref } from 'vue'
import { PropTypes } from '@idux/cdk/utils'
import { IxEmpty } from '@idux/components/empty'
import IxOptionGroup from './OptionGroup.vue'
import IxOption from './Option.vue'
import { visibleChangeToken } from './token'

export default defineComponent({
  name: 'IxOptionContainer',
  components: { IxOptionGroup, IxOption, IxEmpty },
  props: {
    options: PropTypes.array.def(() => []),
    empty: PropTypes.string,
  },
  setup() {
    const showItemCount = ref(0)
    const isEmpty = computed(() => showItemCount.value === 0)

    const visibleChange = (visible: boolean) => {
      if (visible) {
        showItemCount.value++
      } else {
        showItemCount.value--
      }
    }
    provide(visibleChangeToken, visibleChange)

    return { isEmpty }
  },
})
</script>
