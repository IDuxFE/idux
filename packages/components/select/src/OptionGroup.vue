<template>
  <div v-show="visible" class="ix-option-group">
    <span class="ix-option-group-label">
      <slot name="label">{{ label }}</slot>
    </span>
  </div>
  <ix-option v-for="o in options" :key="o.value" :disabled="o.disabled" :label="o.label" :value="o.value" />
  <slot v-if="!options.length" />
</template>
<script lang="ts">
import { computed, defineComponent, inject, nextTick, onUnmounted, provide, ref, watch } from 'vue'
import IxOption from './Option.vue'
import { optionGroupPropsDef } from './types'
import { visibleChangeToken } from './utils'

export default defineComponent({
  name: 'IxOptionGroup',
  components: { IxOption },
  props: optionGroupPropsDef,
  setup() {
    const showItemCount = ref(0)
    const visible = computed(() => showItemCount.value > 0)

    const visibleChange = (visible: boolean) => {
      if (visible) {
        showItemCount.value++
      } else {
        showItemCount.value--
      }
    }
    provide(visibleChangeToken, visibleChange)

    const parentVisibleChange = inject(visibleChangeToken)
    watch(visible, value => nextTick(() => parentVisibleChange?.(value)), { immediate: visible.value })
    onUnmounted(() => {
      if (visible.value) {
        parentVisibleChange?.(false)
      }
    })

    return { visible }
  },
})
</script>
