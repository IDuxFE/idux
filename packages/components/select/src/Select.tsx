/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import { computed, defineComponent, normalizeClass, provide, ref, watch } from 'vue'

import { type VirtualScrollInstance, type VirtualScrollToFn } from '@idux/cdk/scroll'
import { useState } from '@idux/cdk/utils'
import { ɵOverlay } from '@idux/components/_private/overlay'
import { ɵSelector, type ɵSelectorInstance } from '@idux/components/_private/selector'
import { useGlobalConfig } from '@idux/components/config'
import { useFormAccessor } from '@idux/components/utils'

import { useActiveState } from './composables/useActiveState'
import { useFlattedOptions, useMergedOptions } from './composables/useOptions'
import { useOverlayState } from './composables/useOverlayState'
import { useSelectedState } from './composables/useSelectedState'
import Content from './content/Content'
import { selectToken } from './token'
import { selectProps } from './types'

const defaultOffset: [number, number] = [0, 8]

export default defineComponent({
  name: 'IxSelect',
  inheritAttrs: false,
  props: selectProps,
  setup(props, { attrs, expose, slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-select`)
    const config = useGlobalConfig('select')
    const locale = useGlobalConfig('locale')

    const triggerRef = ref<ɵSelectorInstance>()
    const [inputValue, setInputValue] = useState('')
    const focus = () => triggerRef.value?.focus()
    const blur = () => triggerRef.value?.blur()
    const clearInput = () => {
      props.searchable === 'overlay' ? setInputValue('') : triggerRef.value?.clearInput()
    }

    const virtualScrollRef = ref<VirtualScrollInstance>()
    const scrollTo: VirtualScrollToFn = options => virtualScrollRef.value?.scrollTo(options)

    expose({ focus, blur, scrollTo })

    const { overlayRef, overlayStyle, updateOverlay, overlayOpened, setOverlayOpened } = useOverlayState(props)

    const accessor = useFormAccessor()

    const mergedOptions = useMergedOptions(props, slots, config)

    const flattedOptions = useFlattedOptions(props, mergedOptions, inputValue)
    const {
      selectedValue,
      selectedLimit,
      selectedLimitTitle,
      changeSelected,
      selectedOptions,
      handleClear,
      handleItemRemove,
    } = useSelectedState(props, accessor, mergedOptions, locale)

    const { activeIndex, activeOption, changeActive, scrollToActivated } = useActiveState(
      props,
      flattedOptions,
      selectedValue,
      inputValue,
      scrollTo,
    )

    const handleOptionClick = (value: any) => {
      changeSelected(value)
      if (props.multiple) {
        focus()
        clearInput()
      } else {
        setOverlayOpened(false)
      }
    }

    const handleBlur = () => accessor.markAsBlurred()

    provide(selectToken, {
      props,
      slots,
      config,
      mergedPrefixCls,
      virtualScrollRef,
      accessor,
      inputValue,
      setInputValue,
      overlayOpened,
      setOverlayOpened,
      mergedOptions,
      flattedOptions,
      handleOptionClick,
      selectedValue,
      selectedLimit,
      selectedLimitTitle,
      activeIndex,
      activeOption,
      changeActive,
      scrollToActivated,
    })

    watch(overlayOpened, opened => {
      if (!opened && props.allowInput && inputValue.value) {
        changeSelected(inputValue.value)
      }
      opened ? focus() : blur()
      clearInput()
    })

    const overlayClasses = computed(() => {
      const { overlayClassName, multiple } = props
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [`${prefixCls}-overlay`]: true,
        [`${prefixCls}-overlay-multiple`]: multiple,
        [overlayClassName || '']: !!overlayClassName,
      })
    })

    const target = computed(() => props.target ?? config.target ?? `${mergedPrefixCls.value}-overlay-container`)

    const handleKeyDown = (evt: KeyboardEvent) => {
      switch (evt.code) {
        case 'ArrowUp':
          evt.preventDefault()
          changeActive(activeIndex.value - 1, -1)
          break
        case 'ArrowDown':
          evt.preventDefault()
          changeActive(activeIndex.value + 1, 1)
          break
        case 'Enter':
          evt.preventDefault()
          changeSelected(activeOption.value?.value)
          props.multiple ? clearInput() : setOverlayOpened(false)
          break
        case 'Escape':
          evt.preventDefault()
          setOverlayOpened(false)
          break
      }
    }

    const renderTrigger = () => (
      <ɵSelector
        ref={triggerRef}
        v-slots={slots}
        className={mergedPrefixCls.value}
        allowInput={props.allowInput}
        autocomplete={props.autocomplete}
        autofocus={props.autofocus}
        borderless={props.borderless}
        clearable={props.clearable}
        clearIcon={props.clearIcon}
        config={config}
        dataSource={selectedOptions.value}
        disabled={accessor.disabled.value}
        maxLabel={props.maxLabelCount ?? props.maxLabel}
        multiple={props.multiple}
        opened={overlayOpened.value}
        placeholder={props.placeholder}
        readonly={props.readonly}
        searchable={props.searchable}
        size={props.size}
        suffix={props.suffix}
        value={selectedValue.value}
        onBlur={handleBlur}
        onClear={handleClear}
        onInputValueChange={setInputValue}
        onItemRemove={handleItemRemove}
        onKeydown={handleKeyDown}
        onOpenedChange={setOverlayOpened}
        onResize={updateOverlay}
        onSearch={props.onSearch}
        {...attrs}
      />
    )

    const renderContent = () => <Content />

    return () => {
      const overlayProps = {
        class: overlayClasses.value,
        style: overlayStyle.value,
        clickOutside: true,
        disabled: accessor.disabled.value || props.readonly,
        offset: defaultOffset,
        placement: 'bottom',
        target: target.value,
        trigger: 'manual',
        triggerId: attrs.id,
        visible: overlayOpened.value,
        'onUpdate:visible': setOverlayOpened,
      } as const

      const overlaySlots = { default: renderTrigger, content: renderContent }

      return <ɵOverlay ref={overlayRef} {...overlayProps} v-slots={overlaySlots} />
    }
  },
})
