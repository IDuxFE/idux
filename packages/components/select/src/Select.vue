<template>
  <div ref="triggerRef" v-click-outside="onClickOutside" :class="classes" @click="onClick">
    <div class="ix-select-selector" @keydown="onKeyDown">
      <ix-select-item
        v-for="item in selectedItems"
        v-show="showItem"
        :key="item.value"
        :removeable="multiple && !item.isMax"
        :disabled="disabled$$ || item.disabled"
        :label="item.label"
        @delete="onItemDelete(item)"
      >
        <slot v-if="item.isMax" name="customMaxLabel" :option="item"></slot>
        <slot v-else name="customLabel" :option="item"></slot>
      </ix-select-item>
      <ix-select-input
        ref="selectInput"
        :value="inputValue"
        :autofocus="autofocus"
        :disabled="disabled$$"
        :open="open"
        :showInput="multiple || searchable || inputable"
        :showMirror="multiple"
        @change="onInputValueChange"
        @input="$emit('input', $event)"
        @compositionstart="onCompositionStart"
        @compositionend="onCompositionEnd"
        @focus="onFocus"
        @blur="onBlur"
      />
      <div v-if="showPlaceholder" class="ix-select-placeholder">
        <slot name="placeholder">{{ placeholder }}</slot>
      </div>
    </div>
    <div v-if="suffix || $slots.suffix || !multiple" class="ix-select-suffix">
      <slot name="suffix"><ix-icon :name="suffixIcon" /></slot>
    </div>
    <div v-if="clearable && !disabled$$ && selectedItems.length" class="ix-select-clear" @click.stop="onClear">
      <ix-icon name="close-circle" />
    </div>
    <ix-portal target="ix-select-container">
      <transition>
        <ix-option-container
          v-show="visibility"
          ref="overlayRef"
          :class="overlayClass"
          :style="{ minWidth: overlayMinWidth, maxHeight: overlayMaxHeight }"
          :options="normalOptions"
          :empty="empty"
        >
          <slot />
          <template #empty><slot name="empty" /></template>
        </ix-option-container>
      </transition>
    </ix-portal>
  </div>
</template>
<script lang="ts">
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { OptionProps } from './types'

import { computed, defineComponent, ref, watch } from 'vue'
import { clickOutside } from '@idux/cdk/click-outside'
import { IxPortal } from '@idux/cdk/portal'
import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'
import IxSelectItem from './SelectItem.vue'
import IxSelectInput from './SelectInput.vue'
import IxOptionContainer from './OptionContainer.vue'
import { selectProps } from './types'
import {
  useSelectClasses,
  useSelectInput,
  useSelectOptions,
  useSelectOverlay,
  useSelectProvider,
  useSelectValueAccessor,
} from './useSelect'

export default defineComponent({
  name: 'IxSelect',
  components: { IxPortal, IxIcon, IxSelectItem, IxSelectInput, IxOptionContainer },
  directives: { clickOutside },
  props: selectProps,
  emits: [
    'update:value',
    'update:open',
    'blur',
    'change',
    'clear',
    'focus',
    'scrollToBottom',
    'compositionstart',
    'compositionend',
    'input',
    'inputChange',
  ],
  setup(props, { emit }) {
    const config = useGlobalConfig('select')
    const selectedOptions = ref<OptionProps[]>([])

    const { visibility, show, hide, ...overlayBindings } = useSelectOverlay(props)
    const normalOptions = useSelectOptions(props, config)
    const {
      disabled: disabled$$,
      inputValue,
      activatedValue,
      selectedValue,
      isActive,
      ...accessorBindings
    } = useSelectValueAccessor(props)
    const classes = useSelectClasses(props, config, visibility, isActive, disabled$$)
    const { selectInput, focus, blur, clear } = useSelectInput()
    useSelectProvider(props, inputValue, activatedValue, selectedValue, selectedOptions, hide, focus, clear)

    const selectedItems = computed(() => {
      const maxLabelCount = props.maxLabelCount
      const options = selectedOptions.value
      const items: Array<OptionProps & { isMax?: boolean }> = options.slice(0, maxLabelCount)
      if (options.length > maxLabelCount) {
        const label = `+ ${options.length - maxLabelCount} ...`
        const value = options.slice(maxLabelCount)
        items.push({ label, value, disabled: false, isMax: true })
      }
      return items
    })

    const onClickOutside = () => hide()

    const onClick = () => {
      if (disabled$$.value) {
        return
      }
      if (props.multiple) {
        focus()
      }
      if (visibility.value && (props.searchable || props.multiple)) {
        return
      }
      visibility.value ? hide() : show()
    }

    const onKeyDown = (evt: KeyboardEvent) => {
      // TODO support select with directional keyboard
      console.log('Todo onKeyDown', evt)
    }

    watch(visibility, isOpen => {
      if (!isOpen) {
        clear()
      }
      emit('update:open', isOpen)
    })

    const suffixIcon = computed(() => {
      const suffix = props.suffix
      if (suffix) {
        return suffix
      } else {
        return props.searchable && isActive.value ? 'search' : 'down'
      }
    })

    return {
      disabled$$,
      focus,
      blur,
      onClickOutside,
      classes,
      onClick,
      onKeyDown,
      selectedItems,
      selectInput,
      inputValue,
      suffixIcon,
      visibility,
      normalOptions,
      ...overlayBindings,
      ...accessorBindings,
    }
  },
})
</script>
