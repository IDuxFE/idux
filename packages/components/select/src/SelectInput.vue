<template>
  <div class="ix-select-input" :style="{ width: inputWidth }">
    <input
      ref="inputRef"
      class="ix-select-input-inner"
      :value="value"
      autocomplete="off"
      :autofocus="autofocus"
      :disabled="disabled"
      :style="{ opacity: showInput ? null : 0 }"
      @input="onInput"
      @compositionstart="onCompositionStart"
      @compositionend="onCompositionEnd"
      @focus="$emit('focus', $event)"
      @blur="$emit('blur', $event)"
    />
    <span v-if="showMirror" ref="mirrorRef" class="ix-select-input-mirror" aria-hidden="true"></span>
  </div>
</template>
<script lang="ts">
import { defineComponent, onMounted, ref, watch } from 'vue'
import { PropTypes } from '@idux/cdk/utils'

export default defineComponent({
  name: 'IxSelectInput',
  props: {
    value: PropTypes.any,
    autofocus: PropTypes.bool,
    disabled: PropTypes.bool,
    open: PropTypes.bool,
    showInput: PropTypes.bool,
    showMirror: PropTypes.bool,
  },
  emits: ['compositionstart', 'compositionend', 'input', 'change', 'focus', 'blur'],
  setup(props, { emit }) {
    const inputWidth = ref('')
    const inputRef = ref((null as unknown) as HTMLInputElement)
    const mirrorRef = ref((null as unknown) as HTMLInputElement)

    const syncMirrorWidth = () => {
      if (props.showMirror) {
        const inputElement = inputRef.value
        const mirrorElement = mirrorRef.value
        // don't remove the space char, this is placeholder.
        mirrorElement.innerText = ` ${inputElement.value}`
        inputWidth.value = `${mirrorElement.scrollWidth}px`
      }
    }

    const isComposing = ref(false)
    const onCompositionStart = (evt: CompositionEvent) => {
      isComposing.value = true
      emit('compositionstart', evt)
    }
    const onCompositionEnd = (evt: CompositionEvent) => {
      if (isComposing.value) {
        isComposing.value = false
        onInput(evt)
      }
      emit('compositionend', evt)
    }
    const onInput = (evt: Event) => {
      emit('input', evt)
      if (isComposing.value) {
        return
      }
      const { value } = evt.target as HTMLInputElement
      emit('change', value)
      syncMirrorWidth()
    }

    // todo FocusOptions
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const focus = (options?: any) => {
      inputRef.value.focus(options)
    }

    const blur = () => {
      inputRef.value.blur()
    }

    const clear = () => {
      const inputElement = inputRef.value
      inputElement.value = ''
      emit('change', '')

      syncMirrorWidth()
    }

    onMounted(() => {
      watch(
        () => props.open,
        value => value && focus(),
        { immediate: true },
      )

      syncMirrorWidth()
    })

    return { inputWidth, inputRef, mirrorRef, onCompositionStart, onCompositionEnd, onInput, focus, blur, clear }
  },
})
</script>
