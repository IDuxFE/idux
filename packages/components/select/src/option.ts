/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type {
  SelectOptionGroupPublicProps as SelectOptionGroupProps,
  SelectOptionPublicProps as SelectOptionProps,
} from './types'
import type { FunctionalComponent, HTMLAttributes } from 'vue'

const optionKey = '__IDUX_SELECT_OPTION'
const SelectOption = (() => {}) as FunctionalComponent<
  Omit<HTMLAttributes, keyof SelectOptionProps> & SelectOptionProps
>
SelectOption.displayName = 'IxSelectOption'
;(SelectOption as any)[optionKey] = true

const optionGroupKey = '__IDUX_SELECT_OPTION_GROUP'
const SelectOptionGroup = (() => {}) as FunctionalComponent<
  Omit<HTMLAttributes, keyof SelectOptionGroupProps> & SelectOptionGroupProps
>
SelectOptionGroup.displayName = 'IxSelectOptionGroup'
;(SelectOptionGroup as any)[optionGroupKey] = true

export { SelectOption, SelectOptionGroup, optionKey, optionGroupKey }
