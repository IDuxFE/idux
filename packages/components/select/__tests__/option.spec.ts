import { mount, MountingOptions } from '@vue/test-utils'
import IxOption from '../src/Option.vue'
import IxOptionGroup from '../src/OptionGroup.vue'
import { OptionProps, OptionGroupProps } from '../src/types'

describe('Option.vue', () => {
  const OptionMount = (options?: MountingOptions<Partial<OptionProps>>) => mount(IxOption, { ...options })

  afterEach(() => {
    jest.spyOn(console, 'warn').mockClear()
  })

  test('render work', () => {
    const warn = jest.spyOn(console, 'warn').mockImplementation(() => {})
    const wrapper = OptionMount({ props: { label: 'Tom', value: 'tom' } })

    expect(wrapper.html()).toMatchSnapshot()
    expect(warn).toBeCalledTimes(2)
    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.unmount()
    }).not.toThrow()
  })
})

describe('OptionGroup.vue', () => {
  const OptionGroupMount = (options?: MountingOptions<Partial<OptionGroupProps>>) =>
    mount(IxOptionGroup, { ...options })

  afterEach(() => {
    jest.spyOn(console, 'warn').mockClear()
  })

  test('render work', () => {
    const warn = jest.spyOn(console, 'warn').mockImplementation(() => {})
    const wrapper = OptionGroupMount({ props: { label: 'TomGroup' } })

    expect(wrapper.html()).toMatchSnapshot()
    expect(warn).toBeCalledTimes(1)
    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.unmount()
    }).not.toThrow()
  })
})
