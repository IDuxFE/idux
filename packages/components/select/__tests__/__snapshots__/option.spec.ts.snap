// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option.vue render work 1`] = `
"<div class=\\"ix-option\\">
  <!--v-if--><span class=\\"ix-option-label\\">Tom</span>
</div>"
`;

exports[`OptionGroup.vue render work 1`] = `"<div class=\\"ix-option-group\\" style=\\"display: none;\\"><span class=\\"ix-option-group-label\\">TomGroup</span></div>"`;
